ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -214,14 +214,51 @@ protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
   @Override
   void validateKeyPresent(final SourceName sinkName, final Projection projection) {
 
-    final boolean atLeastOneKey = joinKey.getAllViableKeys(schema).stream()
-        .anyMatch(projection::containsExpression);
+    if (joinKey.isForeignKey()) {
+      final DataSource leftInputTable = getLeftmostSourceNode().getDataSource();
+      final SourceName leftInputTableName = leftInputTable.getName();
+      final String plainInputTableName = leftInputTableName.text();
+      final List<Column> leftInputTableKeys = leftInputTable.getSchema().key();
+      final List<Column> missingKeys =
+          leftInputTableKeys.stream().filter(
+              k -> !projection.containsExpression(
+                    new QualifiedColumnReferenceExp(leftInputTableName, k.name()))
+                  && !projection.containsExpression(new UnqualifiedColumnReferenceExp(
+                      ColumnName.of(plainInputTableName + "_" + k.name().text())
+                  ))
+          ).collect(Collectors.toList());
+
+      if (!missingKeys.isEmpty()) {
+        final String columnString = missingKeys.size() == 1 ? "column" : "columns";
+        // column names may be represented as `TABLENAME_COLUMNNAME`
+        final int tablePrefixLength = plainInputTableName.length() + 1;
+        throw new KsqlException(String.format(
+            "Primary key %s missing in projection."
```

Comment:
nit: can we use the same error message format as in `throwKeysNotIncludedError()` for consistency? Ideally we'd just call `throwKeysNotIncludedError()` (and update the method as needed to accommodate this case), or at minimum pull this code out into a new helper method (`throwKeysNotIncludedErrorForFKJoin()` or similar) to make this more readable.

Reply:
I can extract it in a helper, however, the other error message is quite different, given we only handle single-column-keys, and that we might generate a synthetic key column. Might be better to just keep separated for now.